src/compiler/checker.ts

```diff
@@ -31361,6 +31375,28 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         );
     }
 
+    function discriminateContextualTypeByElements(node: ArrayLiteralExpression, contextualType: UnionType) {

```

Comment:
food for thought: `length` should also be treated as a potential discriminant **but** it probably should be ignored when requesting the contextual type for completions. You could try to use `ContextFlags.Completions` for that or you might have to introduce a new context flag for this purpose, smth like `ContextFlags.NoTupleBounds`.

Reply:
removed the length check in `getMatchingUnionConstituentForArrayLiteral` so not sure if this is still relevant

Reply:
it's relevant because this `arg` could contextually be typed as `number`
```ts
const tup: [(arg: number) => void] | [(arg: string) => void, number] = [
  (arg) => {},
];
```